### PR TITLE
torch engine optimize prefill for long context

### PR DESCRIPTION
**Pull request by grimoire**
Long context would move logits to host, which is time consuming. This PR will not output full logits if no request require `return_logits`.

XgIBy2ZTet

`exp` is expensive in cuda (`ex2.approx.f32`). Replace it with `tl.math.fast_expf`
lPab6Db5gB

internlm2_5-7b, tp2, 949840 context length: 808.1619s

### CODE REVIEW

File: lmdeploy/pytorch/engine/engine.py [hunk 1]
```diff
@@ -488,10 +488,10 @@ def async_sampling_logits(self, logits: torch.Tensor,
 
         def __get_last_logits():
             """get last logits."""
-            if inputs.is_decoding:
```
Comment (RunningLeon) [inline lmdeploy/pytorch/engine/engine.py:491]: why change this if statement?

Comment (grimoire) [reply lmdeploy/pytorch/engine/engine.py:491]: if `return_logits` is fails, we will out include logits except the last token.

**Review by RunningLeon — APPROVED** [review]
LGTM

File: lmdeploy/pytorch/engine/engine.py (hunk 1 above)
Comment (grimoire) [reply lmdeploy/pytorch/engine/engine.py:491]: `return_logits=False` will not include non-output logits.

**Comment by lvhan028** [comment]
May merge main so that I can use `profile_generation.py` to benchmark prefill. 
Here is the result from main branch:
```
 python profile_generation.py  /mnt/140/InternLM/internlm2_5-7b-chat-1m/ --tp 4 -c 1 -ct 1 -pt 1000000 --session-len 1048576 --backend pytorch -w 0 -tr 1
profiling ... concurrency: 1, n_prompt_token: 1000000, n_completion_token: 1, test_round: 1, warmup_round: 0
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [29:01<00:00, 1741.85s/it]

--------------------------------------------------
total time: 1741.85s
concurrency: 1, test_round: 1
input_tokens: 1000000, output_tokens: 1
first_token latency(min, max, ave): 1739.192s, 1739.192s, 1739.192s
total_token latency(min, max, ave): 1739.192s, 1739.192s, 1739.192s
token_latency percentiles(50%,75%,95%,99%)(s): [1739.192, 1739.192, 1739.192, 1739.192]
throughput(output): 0.0 token/s
throughput(total): 574.1 token/s
```

**Comment by lvhan028** [comment]
Test 3 rounds:
```
total time: 3925.31s
concurrency: 1, test_round: 3
input_tokens: 1000000, output_tokens: 1
first_token latency(min, max, ave): 1284.825s, 1352.598s, 1307.773s
total_token latency(min, max, ave): 1284.825s, 1352.598s, 1307.773s
token_latency percentiles(50%,75%,95%,99%)(s): [1307.773, 1307.773, 1307.773, 1307.773]
throughput(output): 0.0 token/s
throughput(total): 764.27 token/s
```
From 1741.85 to (1284.825s, 1352.598s, 1307.773) respectively.

**Comment by lvhan028** [comment]
It still got room to be improved. A test result from turbomind engine

```
total time: 431.54s
concurrency: 1, test_round: 1
input_tokens: 1000000, output_tokens: 1
first_token latency(min, max, ave): 431.534s, 431.534s, 431.534s
total_token latency(min, max, ave): 431.534s, 431.534s, 431.534s
token_latency percentiles(50%,75%,95%,99%)(s): [431.534, 431.534, 431.534, 431.534]
throughput(output): 0.0 token/s
throughput(total): 2317.27 token/s
```

**Comment by grimoire** [comment]
```
--------------------------------------------------
total time: 859.87s
concurrency: 1, test_round: 1
input_tokens: 1000000, output_tokens: 1
first_token latency(min, max, ave): 858.838s, 858.838s, 858.838s
total_token latency(min, max, ave): 858.838s, 858.838s, 858.838s
token_latency percentiles(50%,75%,95%,99%)(s): [858.838, 858.838, 858.838, 858.838]
throughput(output): 0.0 token/s
throughput(total): 1162.97 token/s
--------------------------------------------------
```

File: lmdeploy/pytorch/engine/engine_instance.py [hunk 2]
```diff
@@ -42,10 +42,8 @@ async def async_try_add_session(req_sender: RequestSender, session_id: int):
 
 async def async_end(req_sender: RequestSender, session_id: int):
     """End the given session."""
-    resp = await req_sender.async_send(RequestType.END_SESSION,
-                                       dict(session_id=session_id))
-    _check_resp_success(resp, (f'Failed to end session: {session_id}. '
-                               f'Error: {resp.type}.'))
+    await req_sender.async_send_async(RequestType.END_SESSION,
+                                      dict(session_id=session_id, resp=False))
```
Comment (RunningLeon) [inline lmdeploy/pytorch/engine/engine_instance.py:46]: should it be `response=False` as used in https://github.com/InternLM/lmdeploy/blob/5d6c4ab39017d3e1e3e5f2c9edb2f34859e36ee5/lmdeploy/pytorch/engine/engine.py#L248

**Comment by lvhan028** [comment]
I tried the long-context test described in "long_context.md". It didn't generate any token when the `session_len` is 800000. 
But the result shows  "fnish_reason" being  `stop`
```
41221 [Response(text='', generate_token_len=0, input_token_len=767201, session_id=0, finish_reason='stop', token_ids=[], logprobs=None, index=0)]
```

File: lmdeploy/pytorch/engine/engine.py [hunk 3]
```diff
@@ -245,31 +260,37 @@ def _on_add_session(self, reqs: Request, **kwargs):
...
         """on stop session callback."""
         for req in reqs:
             session_id = req.data['session_id']
+            resp = req.data.get('response', True)
             resp_type = ResponseType.SESSION_NOT_EXIST
             if session_id in self.scheduler.sessions:
                 self.scheduler.stop_session(session_id)
                 resp_type = ResponseType.SUCCESS
-            self._response(resp_type, req.sender_id, req.req_id)
+            if resp:
```
Comment (lvhan028) [inline lmdeploy/pytorch/engine/engine.py:280]: There is no "True" case for "_on_stop_session", isn't it?

File: lmdeploy/pytorch/engine/engine.py [hunk 4]
```diff
@@ -245,31 +260,37 @@ def _on_add_session(self, reqs: Request, **kwargs):
...
     def _on_stop_session(self, reqs: Request, **kwargs):
         """on stop session callback."""
         for req in reqs:
             session_id = req.data['session_id']
+            resp = req.data.get('response', True)
             resp_type = ResponseType.SESSION_NOT_EXIST
             if session_id in self.scheduler.sessions:
                 self.scheduler.stop_session(session_id)
                 resp_type = ResponseType.SUCCESS
-            self._response(resp_type, req.sender_id, req.req_id)
```
Comment (lvhan028) [inline lmdeploy/pytorch/engine/engine.py:262]: In `engine_instance.py`, there is no 'response' field when sending "RequestType.STOP_SESSION" data
So, `resp` is always True here, isn't it?

Comment (grimoire) [reply lmdeploy/pytorch/engine/engine.py:262]: It is. We might need no-resp in the future.

**Comment by lvhan028** [comment]
> I tried the long-context test described in "long_context.md". It didn't generate any token when the `session_len` is 800000. But the result shows "fnish_reason" being `stop`
> 
> ```
> 41221 [Response(text='', generate_token_len=0, input_token_len=767201, session_id=0, finish_reason='stop', token_ids=[], logprobs=None, index=0)]
> ```

Forget about this case, I need to use `tp`.  @AllentDan, the `EngineInstance.async_stream_infer` returns an error, but AsyncEngine.generate doesn't handle it. 
```
yield EngineOutput(ResponseType.INPUT_LENGTH_ERROR, [], 0)
```
I think we need to discuss about how to handle the exceptional case

**Comment by AllentDan** [comment]
We can either output the error message or print the error to the console since the server can't be aborted.

**Review by lvhan028 — APPROVED** [review]
